### PR TITLE
Homme: changes in Kokkos.cmake related to internal kokkos build.

### LINKED ISSUE #2964: HOMME: Internal install/use of kokkos buggy
How to reproduce the error:

1) Run cmake, everything works fine. Local install is used, and the flags.cmake of the target (say, sphere_op_ut) has the correct include directory (from the externals/kokkos folder).
2) touch any file that would trigger cmake to be run again (e.g., any CMakeLists.txt) file. Then try to run make. You will get an error like `fatal error: Kokkos_Core.hpp: No such file or directory`. Check the flags.make of the target, and you will find `/usr/local/include` where the kokkos directory was before.

Reason: kokkos defines (at some point) `KOKKOS_PATH to be the installation prefix. Since we don't specify one, the default cmake one is picked (namely, `/usr/local`).

Solution: avoid using cmake variable names that _might_ be used by some tpl in e3sm cmake files. Instead, prepend `E3SM_`, to get a safe variable. In this case, use `E3SM_KOKKOS_PATH` to detect if a Kokkos installation is present.

### PR DESCRIPTION
**Pull request by bartgol**
This fixes issue #2964 . See that issue for details.

### CODE REVIEW

**Comment by jgfouca** [comment]
Thanks for fixing this.

**Comment by ambrad** [comment]
Would a function imp of install_kokkos_if_needed, rather than a macro, give the correct triggering in the case the Kokkos subdirectory is modified (a rare occurrence in any case)? I don't like exposing Kokkos install details in the main CMakeLists.txt.

**Comment by bartgol** [comment]
Interesting. I tried the function, and it works. I re-tried macro, just to double check, and it also works. The fact that it was not working before must have been related to the KOKKOS_PATH being defined rather than the macro itself. I'm going to change that back.

**Comment by bartgol** [comment]
Ok, I changed that back. I can squash the commits into one to keep the history cleaner, but that would require a force push. I can do that once the review is over.

File: components/homme/cmake/Kokkos.cmake [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
-if (NOT DEFINED KOKKOS_PATH)
+if (NOT DEFINED E3SM_KOKKOS_PATH)
   # Build kokkos submodule if user did not specify KOKKOS_PATH.
   set (KOKKOS_INSTALLATION_NEEDED TRUE)
   set (KOKKOS_SRC ${CMAKE_SOURCE_DIR}/../../externals/kokkos)
   set (KOKKOS_INSTALL_DIR ${CMAKE_BINARY_DIR}/kokkos/install)
   SET (KOKKOS_INCLUDE_DIR ${KOKKOS_SRC}/core/src ${KOKKOS_INSTALL_DIR})
   SET (KOKKOS_LIBRARY_DIR ${KOKKOS_INSTALL_DIR})
+  message ("Kokkos will be built internally to E3SM, in ${KOKKOS_INSTALL_DIR}")
+  # SET (E3SM_KOKKOS_PATH ${KOKKOS_INCLUDE_DIR} CACHE PATH "Kokkos installation directory for E3SM")
```
Comment (ambrad) [inline components/homme/cmake/Kokkos.cmake]: Why is there this commented-out, but new, line?

File: components/homme/cmake/Kokkos.cmake [hunk 2]
```diff
@@ -1,12 +1,15 @@
 set (KOKKOS_INSTALLATION_NEEDED FALSE CACHE LOGICAL "")
-if (NOT DEFINED KOKKOS_PATH)
+if (NOT DEFINED E3SM_KOKKOS_PATH)
   # Build kokkos submodule if user did not specify KOKKOS_PATH.
   set (KOKKOS_INSTALLATION_NEEDED TRUE)
   set (KOKKOS_SRC ${CMAKE_SOURCE_DIR}/../../externals/kokkos)
   set (KOKKOS_INSTALL_DIR ${CMAKE_BINARY_DIR}/kokkos/install)
   SET (KOKKOS_INCLUDE_DIR ${KOKKOS_SRC}/core/src ${KOKKOS_INSTALL_DIR})
   SET (KOKKOS_LIBRARY_DIR ${KOKKOS_INSTALL_DIR})
+  message ("Kokkos will be built internally to E3SM, in ${KOKKOS_INSTALL_DIR}")
```
Comment (ambrad) [inline components/homme/cmake/Kokkos.cmake]: "in ${KOKKOS_INSTALL_DIR}" is not quite true. We can't actually get Kokkos to install via the current monolithic add_subdirectory(kokkos) approach. I think it's better not to confuse the user with unnecessary and unusable details, anyway. I recommend you write just the first part: "Kokkos will be built internally to E3SM".

File: components/homme/cmake/Kokkos.cmake (hunk 1 above)
Comment (bartgol) [reply components/homme/cmake/Kokkos.cmake]: Argh. I added it, then commented (it gave the wrong result), then forgot to remove it. I fixed it now.

File: components/homme/cmake/Kokkos.cmake (hunk 2 above)
Comment (bartgol) [reply components/homme/cmake/Kokkos.cmake]: The issue here is the name `KOKKOS_INSTALL_DIR`. Kokkos will actually be built there, but not installed. The printout message is correct though, namely it prints the actual folder where kokkos gets built.

We could change the names given to the cmake variables, namely `KOKKOS_INSTALL_DIR` renamed to `KOKKOS_BUILD_DIR`. And maybe to the folders too (e.g., `${CMAKE_BINARY_DIR}/kokkos/install` should probably be called `${CMAKE_BINARY_DIR}/kokkos/build`). Would you be fine with something like that?

Comment (ambrad) [reply components/homme/cmake/Kokkos.cmake]: OK. Fine as is, or feel free to change. At this point, the PR mechanically works, so it's just a matter of taste.

**Comment by bartgol** [comment]
I force-pushed a squashed version of the commits, since there was no reason to have four commits in the history for changes that add up to less than 10 lines.

**Pull request merged**